### ISSUE TITLE
Issue with unboxability check

### ISSUE THREAD

**Issue by hhugo** [ocaml/ocaml#9624]
The following use to compile with OCaml 4.10 and no longer does (OCaml 4.11).
It was first reported as https://github.com/ocaml/ocaml/issues/9607, The fix in #9609 was not enough and introduced in uncaught exception. It's unclear to me whether there are multiple issues at play here.
 
```ocaml
module Private_variant = struct

  type 'a t = A : 'e -> 'e t

end

module type Public = sig
  module Event : sig
    type 'e t = 'e Private_variant.t
  end
  type t = T : 'e Event.t -> t [@@unboxed]
end

module type Component = sig

  include Public

  module type Public = Public
    with type 'e Event.t = 'e Event.t
     and type t = t
end
```
Here is the error I get with @gasche 's patch from https://github.com/ocaml/ocaml/pull/9609#issuecomment-636342570
```
File "/home/hheuzard/a.ml", line 20, characters 9-19:
20 |      and type t = t
              ^^^^^^^^^^
Error: This type cannot be unboxed because
       it might contain both float and non-float values,
       depending on the instantiation of an unnamed existential variable.
       You should annotate it with [@@ocaml.boxed].
```

@gasche proposed the following slightly simpler repro case
```ocaml
module type Sig = sig
  type 'a ind = 'a * int
  type t = T : 'e ind -> t [@@unboxed]
end

module type Problem = sig
  include Sig
  module type ReboundSig = Sig
    with type 'a ind = 'a ind
     and type t = t
end
```

**Comment by hhugo** [comment]
tentative fix in https://github.com/ocaml/ocaml/pull/9623

**Comment by gasche** [comment]
Here is, from https://github.com/ocaml/ocaml/pull/9623#issuecomment-636860856, a reproduction case for a similar problem with variance instead of separability:

```ocaml
module type Sig = sig
  type +'a abstract
  type +'a user = Foo of 'a abstract
end

module type Problem = sig
  include Sig
  module M : Sig
    with type 'a abstract = 'a abstract
     and type 'a user = 'a user
  type +'a foo = 'a M.user
end
```

```
11 |   type +'a foo = 'a M.user
       ^^^^^^^^^^^^^^^^^^^^^^^^
Error: In this definition, expected parameter variances are not satisfied.
       The 1st type parameter was expected to be covariant,
       but it is injective invariant.
```

**Comment by gasche** [comment]
@hhugo the problem should be fixed now that #9623 is merged (in trunk and 4.11). Would you mind trying to reproduce on your end, to confirm that the issue is fixed?

Note: I proposed an earlier patch, https://github.com/gasche/ocaml/commit/2a142de27c5b1d190a8ad48c43b4f9b4fd7cfef8 , to fix a related issue (an uncaught `Not_found` exception). I currently believe that this patch was not the right way to fix the problem, and it was not included in 4.11. I would recommend removing the patch from your test branch as well, as it could hide other problems. (This first patch was written from the stack trace, without a repro case to play with; it's not surprising but it was much easier to correctly diagnose the issue when you later provided a repro case, thanks again.)

**Comment by hhugo** [comment]
I tested your newer patch two days ago and it was working fine.
I'll rebase again to make sure I have the full diff
I'll close the issue now and reopen if something goes wrong
